internals-js/src/operations.ts

```diff
@@ -653,18 +654,84 @@ export class SelectionSet {
+   *  non frozen sets. Meaning that if one does `s1.mergeIn(s2)` and `s2` is not frozen, then `s1` may (or may not) reference
+   *  `s2` directly (without cloning) and thus later modifications to `s1` may (or may not) modify `s2`. This
+   *  do-not-defensively-clone-by-default behaviour is done for performance reasons.
+   *
+   * Note that freezing is a "deep" operation, in that the whole structure of the selection set is frozen by this method
+   * (and so this is not an excessively cheap operation).
+   *
+   * @return this selection set (for convenience, to allow method chaining).
+   */
+  freeze(): SelectionSet {
+    if (!this.isFrozen()) {
+      for (const selection of this.selections()) {
+        selection.freeze();
+      }
+      this._isFrozen = true;
+    }
+    return this;
+  }
+
+  /**
+   * Whether this selection set is frozen. See `freeze` for details.
+   */
+  isFrozen(): boolean {
+    return this._isFrozen;
+  }
+
+  /**
+   * A shortcut for returning a mutable version of this selection set by cloning it if it is frozen, but returning this set directly
+   * if it is not frozen.
+   */
+  cloneIfFrozen(): SelectionSet {
+    return this.isFrozen() ? this.clone() : this;
+  }
+
+  /**
+   * Adds the selections of the provided selection set to this selection, merging common selection as necessary.
+   *
+   * Please note that by default, the selection from the input may (or may not) be directly referenced by this selection
+   * set after this method return. That is, future modification of this selection set may end up modifying the input
+   * set due to direct aliasing. If direct aliasing should be prevented, the input selection set should be frozen (see
+   * `freeze` for details).
+   */
   mergeIn(selectionSet: SelectionSet) {
     for (const selection of selectionSet.selections()) {
       this.add(selection);
     }
   }
 
+  /**
+   * Adds the provided selections to this selection, merging common selection as necessary.
+   *
+   * This is very similar to `mergeIn` except that it takes a direct array of selection, and the direct aliasing
+   * remarks from `mergeInd` applies here too.
+   */
   addAll(selections: Selection[]): SelectionSet {
     selections.forEach(s => this.add(s));
     return this;
   }
 
+  /**
+   * Adds the provided selection to this selection, merging it to any existing selection of this set as appropriate.
+   *
+   * Please note that by default, the input selection may (or may not) be directly referenced by this selection
+   * set after this method return. That is, future modification of this selection set may end up modifying the input
+   * selection due to direct aliasing. If direct aliasing should be prevented, the input selection should be frozen
+   * (see `freeze` for details).
+   */
   add(selection: Selection): Selection {
+    // It's a but to try to add to a frozen selection set
+    assert(!this.isFrozen(), () => `Cannot add to frozen selection: ${this}`);
```

Comment:
We don't need to add this assertion to other methods? i.e `addPath`, etc?

Reply:
The code make sure all modifications are done by this method, so that one should be sufficient in practice.